src/three-components/Shadow.ts

```diff
@@ -13,26 +13,151 @@
  * limitations under the License.
  */
 
-import {Mesh, PlaneGeometry, ShadowMaterial} from 'three';
+import {Box3, DirectionalLight, Mesh, Object3D, PlaneBufferGeometry, ShadowMaterial, Vector3} from 'three';
+
+import Model from './Model';
+
+// Nothing within Offset of the bottom of the model casts a shadow
+// (this is to avoid having a baked-in shadow plane cast its own shadow).
+const OFFSET = 0.001;
+const BASE_OPACITY = 0.1;
+// The softness [0, 1] of the shadow is mapped to a resolution between
+// 2^LOG_MAX_RESOLUTION and 2^LOG_MIN_RESOLUTION.
+const LOG_MAX_RESOLUTION = 9;
+const LOG_MIN_RESOLUTION = 6;
+// Animated models are not in general contained in their bounding box, as this
+// is calculated only for their resting pose. We create a cubic shadow volume
+// for animated models sized to their largest bounding box dimesion multiplied
+// by this scale factor.
+const ANIMATION_SCALING = 2;
 
 /**
- * Creates a mesh that can receive and render shadows.
+ * The Shadow class creates a shadow that fits a given model and follows a
+ * pivot. Assuming the model is also attached to this pivot, the shadow does not
+ * update due to pivot rotation/translation, provided the setRotation() method
+ * is called whenever the pivot rotates around the Y-axis. We assume the pivot
+ * is restricted to motion in the horizontal plane. The softness of the shadow
+ * is controlled by changing its resolution, making softer shadows faster, but
+ * less precise.
  */
-export default class Shadow extends Mesh {
-  /**
-   * Create a shadow mesh.
-   */
-  constructor() {
-    const geometry = new PlaneGeometry(2000, 2000);
-    geometry.rotateX(-Math.PI / 2);
-
-    const material = new ShadowMaterial(({
-      color: 0x111111,
-      opacity: 0.2,
-    } as any));
-
-    super(geometry, material);
-    this.receiveShadow = true;
-    this.name = 'ShadowMesh';
-  }
-}
+export class Shadow extends DirectionalLight {
+  private shadowMaterial = new ShadowMaterial;
+  private floor: Mesh;
+  private boundingBox = new Box3;
+  private size = new Vector3;
+  public needsUpdate = false;
+
+  constructor(private model: Model, pivot: Object3D, softness: number) {
+    super();
+
+    // We use the light only to cast a shadow, not to light the scene.
+    this.intensity = 0;
+    this.castShadow = true;
+
+    this.floor = new Mesh(new PlaneBufferGeometry, this.shadowMaterial);
+    this.floor.receiveShadow = true;
+    this.floor.castShadow = false;
+
+    pivot.add(this.floor);
+    pivot.add(this);
+    this.target = pivot;
```

Comment:
What if the `Shadow` extended `Object3D`? It could contain both the light and the floor, and any consumer that wishes to add a shadow could roughly follow these steps (psuedocode):

```typescript
class SpecialScene extends Scene {
  constructor() {
    super();
    const model = new Model();
    const shadow = new Shadow(model, softness);
    const pivot = new Object3D();
    pivot.add(model);
    pivot.add(shadow);
    this.add(pivot);
  }
}
```

This would allow us to neatly separate the notions of the pivot from anything contained by the pivot.

Reply:
Glancing down the this file, it looks like both the pivot and the target are not really used beyond basic hierarchy manipulation.

Reply:
Shadow does in fact inherit from Object3D, so the floor part is easy. Since the light needs to set its target to the pivot, I need to include that in the constructor, but I can externalize the rest.